### PR TITLE
Align Test Server direct query behavior with the real server

### LINKED ISSUE #1190: Test Server Query for non-cached execution causes SDK state machines problem
## Actual Behavior
```
io.temporal.client.WorkflowQueryException: workflowId='f2454b12-fcb1-4b3e-9b29-2f02e33a6543', runId='c5e13f9c-0e3c-454f-adb7-e7dd5fea844d', workflowType='TestWorkflowWithQuery'}

	at io.temporal.internal.sync.WorkflowStubImpl.query(WorkflowStubImpl.java:331)
	at io.temporal.testing.TimeLockingInterceptor$TimeLockingWorkflowStub.query(TimeLockingInterceptor.java:172)
	at io.temporal.internal.sync.WorkflowInvocationHandler$SyncWorkflowInvocationHandler.queryWorkflow(WorkflowInvocationHandler.java:309)
	at io.temporal.internal.sync.WorkflowInvocationHandler$SyncWorkflowInvocationHandler.invoke(WorkflowInvocationHandler.java:272)
	at io.temporal.internal.sync.WorkflowInvocationHandler.invoke(WorkflowInvocationHandler.java:178)
	at com.sun.proxy.$Proxy18.query(Unknown Source)
	at io.temporal.workflow.queryTests.LocalActivityAndQueryInvestigationTest.testLocalActivityAndQuery(LocalActivityAndQueryInvestigationTest.java:68)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: io.temporal.internal.replay.InternalWorkflowTaskException: Failure handling event 5 of type 'EVENT_TYPE_MARKER_RECORDED' during execution. {PreviousStartedEventId=3, workflowTaskStartedEventId=3, Currently Processing StartedEventId=3}
	at io.temporal.internal.statemachines.WorkflowStateMachines.createEventProcessingException(WorkflowStateMachines.java:244)
	at io.temporal.internal.statemachines.WorkflowStateMachines.handleEventsBatch(WorkflowStateMachines.java:223)
	at io.temporal.internal.statemachines.WorkflowStateMachines.handleEvent(WorkflowStateMachines.java:197)
	at io.temporal.internal.replay.ReplayWorkflowRunTaskHandler.handleWorkflowTaskImpl(ReplayWorkflowRunTaskHandler.java:180)
	at io.temporal.internal.replay.ReplayWorkflowRunTaskHandler.handleQueryWorkflowTask(ReplayWorkflowRunTaskHandler.java:247)
	at io.temporal.internal.replay.ReplayWorkflowTaskHandler.handleWorkflowTaskWithQuery(ReplayWorkflowTaskHandler.java:118)
	at io.temporal.internal.replay.ReplayWorkflowTaskHandler.handleWorkflowTask(ReplayWorkflowTaskHandler.java:97)
	at io.temporal.internal.worker.WorkflowWorker$TaskHandlerImpl.handleTask(WorkflowWorker.java:277)
	at io.temporal.internal.worker.WorkflowWorker$TaskHandlerImpl.handle(WorkflowWorker.java:231)
	at io.temporal.internal.worker.WorkflowWorker$TaskHandlerImpl.handle(WorkflowWorker.java:173)
	at io.temporal.internal.worker.PollTaskExecutor.lambda$process$0(PollTaskExecutor.java:93)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalArgumentException: LocalActivity: invalid REQUEST_PREPARED->MARKER_RECORDED, transition history is [CREATED->CHECK_EXECUTION_STATE, EXECUTING->SCHEDULE]
	at io.temporal.internal.statemachines.StateMachine.executeTransition(StateMachine.java:140)
	at io.temporal.internal.statemachines.StateMachine.handleHistoryEvent(StateMachine.java:101)
	at io.temporal.internal.statemachines.EntityStateMachineBase.handleEvent(EntityStateMachineBase.java:67)
	at io.temporal.internal.statemachines.WorkflowStateMachines.handleLocalActivityMarker(WorkflowStateMachines.java:482)
	at io.temporal.internal.statemachines.WorkflowStateMachines.handleCommandEvent(WorkflowStateMachines.java:292)
	at io.temporal.internal.statemachines.WorkflowStateMachines.handleSingleEvent(WorkflowStateMachines.java:250)
	at io.temporal.internal.statemachines.WorkflowStateMachines.handleEventsBatch(WorkflowStateMachines.java:221)
	... 12 more

	at io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:271)
	at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:252)
	at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:165)
	at io.temporal.api.workflowservice.v1.WorkflowServiceGrpc$WorkflowServiceBlockingStub.queryWorkflow(WorkflowServiceGrpc.java:3135)
	at io.temporal.internal.client.external.GenericWorkflowClientExternalImpl.lambda$query$5(GenericWorkflowClientExternalImpl.java:220)
	at io.temporal.internal.retryer.GrpcSyncRetryer.retry(GrpcSyncRetryer.java:61)
	at io.temporal.internal.retryer.GrpcRetryer.retryWithResult(GrpcRetryer.java:51)
	at io.temporal.internal.client.external.GenericWorkflowClientExternalImpl.query(GenericWorkflowClientExternalImpl.java:213)
	at io.temporal.internal.client.RootWorkflowClientInvoker.query(RootWorkflowClientInvoker.java:141)
	at io.temporal.internal.sync.WorkflowStubImpl.query(WorkflowStubImpl.java:324)
	... 18 more
```

## Root cause

The way Test Server forms a `PollWorkflowTaskResponse` for a legacy query during replay makes SDK state machines think that the last workflow task is actually getting executed for the first time instead of being replayed (`previousStartedEventId` and `startedEventId` have the same values as the last executed workflow task) causing a failure when SDK tries to actually match already existing events with commands. This issue appears specifically with the LocalActivity state machine.

### PR DESCRIPTION
**Pull request by Spikhalskiy**
## What was changed

Add a unit test for a specific corner case of a direct query after completion of local activity.
Align Test Server behavior with real Temporal Server for such a direct query.

## Why?

The current unexpected behavior of Temporal Test Server causes state machine problems.

Closes #1190

### CODE REVIEW

File: temporal-testing/src/main/java/io/temporal/testing/internal/SDKTestWorkflowRule.java [hunk 1]
```diff
@@ -299,6 +307,31 @@ public static void assertNoHistoryEvent(History history, EventType eventType) {
     }
   }
 
+  /** Waits till the end of the workflow task if there is a workflow task in progress */
+  public void waitForTheEndOfWFT(WorkflowExecution execution) {
+    WorkflowExecutionHistory initialHistory = getExecutionHistory(execution);
+
+    HistoryEvent lastEvent = initialHistory.getLastEvent();
+    if (isWFTInProgress(lastEvent)) {
```
Comment (Sushisource) [inline temporal-testing/src/main/java/io/temporal/testing/internal/SDKTestWorkflowRule.java:315]: Could this loop possibly skip over a whole WFT and think it's still "in progress" because now we're on to the next one?

Would it be reasonable to use some kind of wake-up rather than the busy sleep?

**Review by Sushisource — APPROVED** [review]
Makes sense to me, would be good if we could eliminate those busy sleeps somehow but I get it if it's a huge chunk of work

File: temporal-testing/src/main/java/io/temporal/testing/internal/SDKTestWorkflowRule.java (hunk 1 above)
Comment (Spikhalskiy) [reply temporal-testing/src/main/java/io/temporal/testing/internal/SDKTestWorkflowRule.java:315]: > Could this loop possibly skip over a whole WFT and think it's still "in progress" because now we're on to the next one?        
```
        for (previous = historyEventListIterator.previous();
            historyEventListIterator.hasPrevious() && previous.getEventId() > startEventId;
            previous = historyEventListIterator.previous()) {
          if (!isWFTInProgress(historyEventListIterator.previous())) {
            return;
          }
        }
```
No, it's not. We remember where we start and we check the last position up to the started position trying to sign something other than SCHEDULED and STARTED. If we find - we return. So, if the new task will be added and is in STARTED position for example - we will catch it.

> Would it be reasonable to use some kind of wake-up rather than the busy sleep?
It 100% will be if it's an SDK code. For test purposes busy waiting is fine I think.
